PVTPackage/source/MultiphaseSystem/PhaseSplitModel/DeadOilFlash.cpp

```diff
@@ -54,23 +55,46 @@ BlackOil_WaterModel const & DeadOilFlash::getWaterPhaseModel() const
 
 bool DeadOilFlash::computeEquilibrium( DeadOilFlashMultiphaseSystemProperties & sysProps ) const
 {
+  // OIL  
   const auto & pressure = sysProps.getPressure();
 
-  const DeadOil_PhaseModel & oilPhaseModel = getOilPhaseModel();
-  const DeadOil_PhaseModel & gasPhaseModel = getGasPhaseModel();
-  const BlackOil_WaterModel & waterPhaseModel = getWaterPhaseModel();
-
-  // OIL
+  const DeadOil_PhaseModel & oilPhaseModel = getOilPhaseModel();  
   auto const oilProps = oilPhaseModel.computeProperties( pressure );
   sysProps.setOilModelProperties( oilProps );
 
-  // GAS
-  auto const gasProps = gasPhaseModel.computeProperties( pressure );
-  sysProps.setGasModelProperties( gasProps );
-
-  // WATER
-  auto const waterProps = waterPhaseModel.computeProperties( pressure );
-  sysProps.setWaterModelProperties( waterProps );
+  size_t const nPhases = sysProps.getPhases().size();
+
+  if( nPhases == 3 )
```

Comment:
What about something like
```c++
bool const containsGas = ...
bool const containsWater = ...
if( containsGas == containsWater ) { SomethingWrongIsHappening(); }
```
on top (probably need to factorize a convenient common business interface here) and then
```c++
const DeadOil_PhaseModel & oilPhaseModel = getOilPhaseModel();  
auto const oilProps = oilPhaseModel.computeProperties( pressure );
sysProps.setOilModelProperties( oilProps );

if( containsGas )
{
    const DeadOil_PhaseModel & gasPhaseModel = getGasPhaseModel();
    auto const gasProps = gasPhaseModel.computeProperties( pressure );
    sysProps.setGasModelProperties( gasProps );
}

if( containsWater )
{
      const BlackOil_WaterModel & waterPhaseModel = getWaterPhaseModel();
      auto const waterProps = waterPhaseModel.computeProperties( pressure );
      sysProps.setWaterModelProperties( waterProps );
}
```
so the we keep it non duplicated and we do not mix the different logics.

Reply:
There may also be a refactored solution like
```c++
for( const pvt::PHASE_TYPE & phase: sysProps.getPhases() ) {
  const auto & phaseModel = getPhaseModel( phase );
  const auto props = phaseModel.computeProperties( pressure );
  sysProps.setModelProperties( phase, props );
}
```
but I'm not sure it can work smoothly since `DeadOil_PhaseModel` and `BlackOil_WaterModel` have nothing in common (IIRC).

Reply:
I implemented the first option that you suggested. 

That being said, what I did in this PR introduces some indirection in the DO PVT calculations. It would be nice to find a way to tell `PVTPackage` how many phases are present and which phases they are before launching the kernel update on all the cells (for instance with a template?). Then we could have a specialization of this function (and all the functions called during the update that contain some indirection) to handle all the 3 possible cases without any `if` statement.

For a future refactoring, I have an off-topic question regarding this function but I am not sure so let me know what you think about it. In this function, we involve a lot of classes (`DeadOil_PhaseModel`, `BlackOil_WaterPhase`, `BlackOilDeadOilSystemProperties`, `DeadOilFlash`) with getters and setters to access their data members, but in fact, what we do is quite simple: we fill the maps `massDensity`, `moleDensity`, `molecularWeight`, and `viscosity` and we return. For `BlackOil` and compositional models, more arrays would be updated (`phaseFractions`, `phaseCompFractions`, etc) but the principle remains essentially the same with more complexity (we would actually compute an equilibrium).

So we could do the following with just an in-out function (containing possibly nested in-out functions for modularity) that does not call any class data member:
```
bool DeadOilFlash::computeEquilibrium( real64 const pressure, std::vector< real64 > & massDensity,  std::vector< real64 > & moleDensity, ...  ) const
{
   massDensity[oilPhaseId] = ...; // use alg. currently in DeadOil_PhaseModel
   massDensity[waterPhaseId] = ...; // use alg. currently in BlackOil_WaterModel
  
   viscosity[oilPhaseId] = ...;
   viscosity[waterPhaseId] = ...;

   ... update other properties as needed
}
```
in which we would simply fill the arrays ( `std::vector< real64 >` for now) that we need to fill and return. Later we could use some flags or templates to specialize the function depending on the number of phases, etc. We would not need to call any getter or setter. Ultimately if we propagate this logic everywhere, maybe we would not even need `MultiPhaseSystemProperties` or the `PhaseModels`, and we would just have stateless functions taking everything they need for input/output as arguments. Do you think that could simplify things (just disregard this comment if it does not make sense :) )?